File: API/PetCenterModels/DataTransferObjects/CategoryDTO.cs
using Microsoft.VisualBasic;
using PetCenterModels.DBTables;
using PetCenterModels.DataTransferObjects;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;
using System.Data.Common;

namespace PetCenterModels.DataTransferObjects
{

    public class UsageSubDTO : IBaseResponseDTO<Usage, UsageSubDTO>
    {
        public Guid? Id {get; set;}

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public Guid CategoryId {get; set;}
        public Guid KindId {get; set;}

        public AnimalScale? ScaleSpecific {get; set;} = null;

        public int AverageDailyAmountGrams {get; set;}
        

        public static UsageSubDTO? FromEntity(Usage? usage)
        {
            if(usage==null){return null;}
            return new UsageSubDTO
            {
                Id=usage.Id,
                CurrentVersion=usage.CurrentVersion,
                CategoryId=usage.CategoryId,
                KindId=usage.KindId,
                ScaleSpecific=usage.ScaleSpecific,
                AverageDailyAmountGrams=usage.AverageDailyAmountGrams
            };

        }


    }
    public class CategoryDTO : ISerializableRequestDTO<Category>, IBaseResponseDTO<Category,CategoryDTO>
    {
       
        public Guid? Id {get; set;} = null;

        public byte[] CurrentVersion { get; set; } = Array.Empty<byte>();

        public List<NoteSubDTO>? Notes {get; set;} = null;

        public string Title {get; set;} = string.Empty;

        public bool Consumable { get; set; } = false;

        public List<UsageSubDTO?>? UsageSpecifics {get; set;} = null;

        public static CategoryDTO? FromEntity(Category? entity)
        {
            if(entity==null){return null;}
            return new CategoryDTO
            {
                Id = entity.Id,
                CurrentVersion= entity.CurrentVersion,
                Title = entity.Title,
                Consumable = entity.Consumable,
                UsageSpecifics = entity.UsageSpecifics.Select(u=>UsageSubDTO.FromEntity(u)).ToList()
            };
        }

        public Category? ToEntity()
        {
            Category category = new();
            category.CurrentVersion=CurrentVersion;
            category.Title=Title;
            category.Consumable=Consumable;
            return category;
        }
        
        
        public bool Validate()
        {
            return !string.IsNullOrWhiteSpace(Title);
        }



    }
}

File: Frontend/pet_center_app/lib/models/data_transfer/category_dto.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:pet_center_app/models/data_transfer/note_sub_dto.dart';
import 'package:pet_center_app/models/enums.dart';
part 'category_dto.g.dart';

@JsonSerializable(explicitToJson: true)
class UsageSubDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String categoryId;
  String kindId;
  AnimalScale? scaleSpecific;
  int averageDailyAmountGrams;

  UsageSubDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.categoryId = '',
    this.kindId = '',
    this.scaleSpecific,
    this.averageDailyAmountGrams = 0,
  });

  UsageSubDTO copy() => UsageSubDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    categoryId: categoryId,
    kindId: kindId,
    scaleSpecific: scaleSpecific,
    averageDailyAmountGrams: averageDailyAmountGrams,
  );

  factory UsageSubDTO.fromJson(Map<String, dynamic> json) =>
      _$UsageSubDTOFromJson(json);
  Map<String, dynamic> toJson() => _$UsageSubDTOToJson(this);
}

@JsonSerializable(explicitToJson: true)
class CategoryDTO {
  String? id;
  String currentVersion;
  List<NoteSubDTO>? notes;
  String title;
  bool consumable;
  List<UsageSubDTO>? usageSpecifics;

  CategoryDTO({
    this.id,
    this.currentVersion = '',
    this.notes,
    this.title = '',
    this.consumable = false,
    this.usageSpecifics,
  });

  CategoryDTO copy() => CategoryDTO(
    id: id,
    currentVersion: currentVersion,
    notes: notes?.map((n) => n.copy()).toList(),
    title: title,
    consumable: consumable,
    usageSpecifics: usageSpecifics?.map((u) => u.copy()).toList(),
  );

  factory CategoryDTO.fromJson(Map<String, dynamic> json) =>
      _$CategoryDTOFromJson(json);
  Map<String, dynamic> toJson() => _$CategoryDTOToJson(this);
}
